x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_save.ts

```diff
@@ -0,0 +1,297 @@
+import { refreshTimelines } from './helpers';
+
+function isSaveTimelineAction(action: Action): action is ReturnType<typeof saveTimeline> {
+  return action.type === saveTimeline.type;
+}
+
+export const saveTimelineMiddleware: (kibana: CoreStart) => Middleware<{}, State> =
+  (kibana: CoreStart) => (store) => (next) => async (action: Action) => {
+    // perform the action
+    const ret = next(action);
+
+    if (isSaveTimelineAction(action)) {
+      const { id: localTimelineId } = action.payload;
+      const timeline = selectTimelineById(store.getState(), localTimelineId);
+      const { timelineId, timelineVersion, templateTimelineId, templateTimelineVersion } =
+        extractTimelineIdsAndVersions(timeline);
+      const timelineTimeRange = inputsSelectors.timelineTimeRangeSelector(store.getState());
+
+      store.dispatch(startTimelineSaving({ id: localTimelineId }));
+
+      try {
+        const result = await (action.payload.saveAsNew && timeline.id
+          ? copyTimeline({
+              timelineId,
+              timeline: {
+                ...convertTimelineAsInput(timeline, timelineTimeRange),
+                templateTimelineId,
+                templateTimelineVersion,
+              },
+              savedSearch: timeline.savedSearch,
+            })
+          : persistTimeline({
+              timelineId,
+              version: timelineVersion,
+              timeline: {
+                ...convertTimelineAsInput(timeline, timelineTimeRange),
+                templateTimelineId,
+                templateTimelineVersion,
+              },
+              savedSearch: timeline.savedSearch,
+            }));
+
+        if (isTimelineErrorResponse(result)) {
+          const error = getErrorFromResponse(result);
+          switch (error?.errorCode) {
+            // conflict
+            case 409:
+              kibana.notifications.toasts.addDanger({
+                title: i18n.TIMELINE_VERSION_CONFLICT_TITLE,
+                text: i18n.TIMELINE_VERSION_CONFLICT_DESCRIPTION,
+              });
+              break;
+            default:
+              kibana.notifications.toasts.addDanger({
+                title: i18n.UPDATE_TIMELINE_ERROR_TITLE,
+                text: error?.message ?? i18n.UPDATE_TIMELINE_ERROR_TEXT,
+              });
+          }
+          return;
+        }
+
+        const response = result.data.persistTimeline;
+        if (response == null) {
+          kibana.notifications.toasts.addDanger({
+            title: i18n.UPDATE_TIMELINE_ERROR_TITLE,
+            text: i18n.UPDATE_TIMELINE_ERROR_TEXT,
+          });
+          return;
+        }
+
+        if (response && response.code === 403) {
+          store.dispatch(showCallOutUnauthorizedMsg());
+          return;
+        }
+
+        refreshTimelines(store.getState());
+
+        store.dispatch(
+          updateTimeline({
+            id: localTimelineId,
+            timeline: {
```

Comment:
Thought, for a future reference. If we have the timeline shape stored in zod, we could just use `zod.parse()` on the response and spread that and not worry about missing a value or spreading something incorrectly. Thoughts?

Reply:
Sounds good to me. We would just need to make sure to also update only-local timeline fields (such as `isSaving`)